### PR TITLE
ENH Adding support for NetworkX Graphs as inputs [REVIEW]

### PR DESCRIPTION
**Pull request by BradReesWork**
This PR starts adding support for passing in a NetworkX graph as input to cuGraph algorithms.

import networkx as nx
import cugraph as cuda.nx

... create Nx graph ...  Gnx
>>  pr = nx.pagerank(Gnx)
 that becomes
>>  pr = cuda.nx.pagerank(Gnx)

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by BradReesWork** [comment]
rerun tests

**Comment by BradReesWork** [comment]
rerun tests

**Comment by BradReesWork** [comment]
rerun tests

File: python/cugraph/centrality/betweenness_centrality.py [hunk 1]
```diff
@@ -197,7 +205,7 @@ def edge_betweenness_centrality(
 
     Returns
     -------
-    df : cudf.DataFrame
+    df : cudf.DataFrame or dictionary
```
Comment (Iroy30) [inline python/cugraph/centrality/betweenness_centrality.py:208]: better if it is consistent with betweenness_centrality doc "df : cudf.DataFrame or Dictionary if using NetworkX"

File: python/cugraph/centrality/katz_centrality.py [hunk 2]
```diff
@@ -12,6 +12,7 @@
 # limitations under the License.
 
 from cugraph.centrality import katz_centrality_wrapper
+import cugraph
 
 
 def katz_centrality(
```
Comment (Iroy30) [inline python/cugraph/centrality/katz_centrality.py:18]: doc update for nx

File: python/cugraph/community/ktruss_subgraph.py [hunk 3]
```diff
@@ -13,6 +13,43 @@
...
+        The desired k to be used for extracting the k-truss subgraph.
+
+    Returns
+    -------
+    G_truss : cuGraph.Graph or networkx.Graph
+        A cugraph graph descriptor with the k-truss subgraph for the given k.
+        The networkx graph will NOT have all attributes copied over
+    """
+
+    is_nx = is_networkx_graph(G)
```
Comment (Iroy30) [inline python/cugraph/community/ktruss_subgraph.py:44]: would it be better to use check_nx_graph to be consistent?

Comment (BradReesWork) [reply python/cugraph/community/ktruss_subgraph.py:44]: the issue was that I needed to keep a reference to the original input Graph. But I could uuse nx_check_graph and just update variable names

File: python/cugraph/centrality/betweenness_centrality.py (hunk 1 above)
Comment (BradReesWork) [reply python/cugraph/centrality/betweenness_centrality.py:208]: updated

File: python/cugraph/centrality/katz_centrality.py (hunk 2 above)
Comment (BradReesWork) [reply python/cugraph/centrality/katz_centrality.py:18]: fixed

**Comment by BradReesWork** [comment]
rerun tests

**Comment by BradReesWork** [comment]
rerun tests

**Review by Iroy30 — APPROVED** [review]

File: python/cugraph/link_analysis/pagerank.py [hunk 4]
```diff
@@ -13,10 +13,12 @@
 
 from cugraph.link_analysis import pagerank_wrapper
 from cugraph.structure.graph import null_check
+import cugraph
 
 
 def pagerank(
-    G, alpha=0.85, personalization=None, max_iter=100, tol=1.0e-5, nstart=None
+    G, alpha=0.85, personalization=None, max_iter=100, tol=1.0e-5, nstart=None,
```
Comment (rlratzel) [inline python/cugraph/link_analysis/pagerank.py:20]: Nx default tolerance is 1e-8, does that matter?

File: python/cugraph/tests/test_balanced_cut.py [hunk 5]
```diff
@@ -120,3 +121,39 @@ def test_digraph_rejected():
...
     with pytest.raises(Exception):
         cugraph_call(G, 2)
+
+
+@pytest.mark.parametrize("graph_file", utils.DATASETS)
+@pytest.mark.parametrize("partitions", PARTITIONS)
+def test_edge_cut_clustering_with_edgevals_nx(graph_file, partitions):
+    gc.collect()
+
+    # Read in the graph and get a cugraph object
```
Comment (rlratzel) [inline python/cugraph/tests/test_balanced_cut.py:131]: "...and get a Nx graph object"?

File: python/cugraph/tests/test_balanced_cut.py [hunk 6]
```diff
@@ -120,3 +121,39 @@ def test_digraph_rejected():
...
+    NM = utils.read_csv_for_nx(graph_file, read_weights_in_sp=True)
+    G = nx.from_pandas_edgelist(
+                NM, create_using=nx.Graph(), source="0", target="1",
+                edge_attr="weight"
+    )
```
Comment (rlratzel) [inline python/cugraph/tests/test_balanced_cut.py:136]: could you have used `utils.generate_nx_graph_from_file()` here instead?

File: python/cugraph/tests/test_connectivity.py [hunk 7]
```diff
@@ -205,3 +190,45 @@ def test_strong_cc(graph_file):
 
     cg_vertices = sorted(lst_cg_components[idx])
     assert nx_vertices == cg_vertices
+
+
+# Test all combinations of default/managed and pooled/non-pooled allocation
```
Comment (rlratzel) [inline python/cugraph/tests/test_connectivity.py:195]: Is this a comment a FIXME? If not, then I don't think that is happening.

**Review by rlratzel — APPROVED** [review]
I didn't see any bugs but did have a few optional suggestions which should not hold up my approval and can be addressed in a future PR.

**Comment by BradReesWork** [comment]
rerun tests

**Comment by BradReesWork** [comment]
rerun tests

**Comment by BradReesWork** [comment]
rerun tests

File: python/cugraph/centrality/betweenness_centrality.py [hunk 8]
```diff
@@ -91,7 +93,7 @@ def betweenness_centrality(
 
     Returns
     -------
-    df : cudf.DataFrame
+    df : cudf.DataFrame or Dictionary if using NetworkX
         GPU data frame containing two cudf.Series of size V: the vertex
```
[the remark this answers is not in the document]
Comment (BradReesWork) [reply python/cugraph/centrality/betweenness_centrality.py:97]: fixed

File: python/cugraph/centrality/betweenness_centrality.py [hunk 9]
```diff
@@ -42,7 +44,7 @@ def betweenness_centrality(
 
     Parameters
     ----------
-    G : cuGraph.Graph
+    G : cuGraph.Graph or networkx.Graph
         cuGraph graph descriptor with connectivity information. The graph can
```
[the remark this answers is not in the document]
Comment (BradReesWork) [reply python/cugraph/centrality/betweenness_centrality.py:48]: fixed

File: python/cugraph/community/ecg.py [hunk 10]
```diff
@@ -41,6 +43,10 @@ def ecg(input_graph, min_weight=0.05, ensemble_size=16):
         The default value is 16, larger values may produce higher quality
         partitions for some graphs.
 
+    weight : str
+        Edge data column to use.  Default is None
+        This parameter is here for NetworkX compatibility
```
[the remark this answers is not in the document]
Comment (BradReesWork) [reply python/cugraph/community/ecg.py:48]: fixed text

File: python/cugraph/community/ktruss_subgraph.py [hunk 11]
```diff
@@ -13,6 +13,41 @@
...
 from cugraph.structure.graph import Graph
+from cugraph.utilities import check_nx_graph
+from cugraph.utilities import cugraph_to_nx
+
+
+def k_truss(G, k):
+    """
+    Returns the K-Truss subgraph of a graph for a specific k.
+
+    This API is for NwtworkX compatability.
```
[the remark this answers is not in the document]
Comment (BradReesWork) [reply python/cugraph/community/ktruss_subgraph.py:24]: fixed

File: python/cugraph/utilities/nx_factory.py [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    del pdf
+
+    return G
+
+
+def is_networkx_graph(G):
+    return isinstance(G, nx.classes.graph.Graph)
+
+
+def check_nx_graph(G, weight=None):
```
[the remark this answers is not in the document]
Comment (BradReesWork) [reply python/cugraph/utilities/nx_factory.py:57]: I agree the name could be better, and should be.  The function checks and converts Nx Graphs to cuGraph.  So "ensure" might also not be the right name

File: python/cugraph/link_analysis/pagerank.py [hunk 13]
```diff
@@ -67,6 +69,13 @@ def pagerank(
         nstart['values'] : cudf.Series
             Pagerank values for vertices
 
+    weight : str
+        Edge data column to use.  Default is None
+        This version of PageRank current does not use edge weight.
+        This parameter is here for NetworkX compatibility
+    dangling : dict
+        This parameter is here for NetworkX compatibility
```
[the remark this answers is not in the document]
Comment (BradReesWork) [reply python/cugraph/link_analysis/pagerank.py:77]: fixed

File: python/cugraph/link_analysis/pagerank.py (hunk 4 above)
Comment (BradReesWork) [reply python/cugraph/link_analysis/pagerank.py:20]: I need to run a set of benchmarks to determine if there is an benefit to the the lower value

File: python/cugraph/tests/test_balanced_cut.py (hunk 5 above)
Comment (BradReesWork) [reply python/cugraph/tests/test_balanced_cut.py:131]: fixed

File: python/cugraph/tests/test_balanced_cut.py (hunk 6 above)
Comment (BradReesWork) [reply python/cugraph/tests/test_balanced_cut.py:136]: true.  I should update all the test to use that

File: python/cugraph/tests/test_connectivity.py (hunk 7 above)
Comment (BradReesWork) [reply python/cugraph/tests/test_connectivity.py]: fixed

**Pull request merged**